src/compiler/checker.ts

```diff
@@ -9526,6 +9526,17 @@ namespace ts {
             return type.flags & TypeFlags.IndexedAccess ? getSimplifiedIndexedAccessType(<IndexedAccessType>type) : type;
         }
 
+        function distributeIndexOverObjectType(objectType: Type, indexType: Type) {
+            // (T | U)[K] -> T[K] | U[K]
+            if (objectType.flags & TypeFlags.Union) {

```

Comment:
You're a bad person for not testing this branch.

Reply:
Excuse me, I'm just reusing an operation already in use elsehwere ☑️